File: app/src/main/java/de/tum/in/tca/ticketcheck/component/ticket/activity/AdminDetailsActivity.java
package de.tum.in.tca.ticketcheck.component.ticket.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.widget.SearchView;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import de.tum.in.tca.ticketcheck.R;
import de.tum.in.tca.ticketcheck.api.TUMCabeClient;
import de.tum.in.tca.ticketcheck.component.generic.activity.BaseActivity;
import de.tum.in.tca.ticketcheck.component.ticket.TicketsController;
import de.tum.in.tca.ticketcheck.component.ticket.model.AdminTicket;
import de.tum.in.tca.ticketcheck.component.ticket.payload.TicketStatus;
import de.tum.in.tca.ticketcheck.database.TcaDb;
import de.tum.in.tca.ticketcheck.utils.Utils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class AdminDetailsActivity extends BaseActivity {

    private SearchView searchView;
    private ListView listView;
    private TextView totalSaleView;
    private TicketListAdapter ticketListAdapter;
    private List<AdminTicket> tickets;

    private int eventID;
    private int totalTicketCount;
    private TicketsController ticketsController;
    private int lastSelectedIndex;

    public AdminDetailsActivity() {
        super(R.layout.activity_admin);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        searchView = findViewById(R.id.search_ticket);
        totalSaleView = findViewById(R.id.sold_tickets);
        listView = findViewById(R.id.ticket_list);
        FloatingActionButton floatingScanner = findViewById(R.id.fab_scanner);

        eventID = getIntent().getIntExtra("event_id", 0);

        ticketsController = new TicketsController(this);
        // Retrieve the saved ticket data from local database
        tickets = ticketsController.getTicketsForEvent(eventID);
        ticketListAdapter = new TicketListAdapter(tickets, AdminDetailsActivity.this);
        listView.setAdapter(ticketListAdapter);
        ticketListAdapter.notifyDataSetChanged();
        listView.setOnItemClickListener(new android.widget.AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                AdminTicket ticket = (AdminTicket) parent.getItemAtPosition(position);
                lastSelectedIndex = position;
                Intent intent = new Intent(AdminDetailsActivity.this, TicketDetailsActivity.class);
                intent.putExtra("ticketId", ticket.getId());
                startActivity(intent);
            }
        });

        // Invoke ticket refresh
        refreshTicketList();

        //Set total ticket and total sale
        totalTicketCount = 0;
        updateTicketCounter();
        TUMCabeClient.getInstance(AdminDetailsActivity.this).getTicketStats(eventID, new Callback<List<TicketStatus>>() {
            @Override
            public void onResponse(@NonNull Call<List<TicketStatus>> call, @NonNull Response<List<TicketStatus>> response) {
                List<TicketStatus> list = response.body() != null
                        ? response.body()
                        : new ArrayList<>();
                totalTicketCount = 0;
                for (TicketStatus status : list) {
                    totalTicketCount += status.getContingent();
                }
                updateTicketCounter();
            }

            @Override
            public void onFailure(@NonNull Call<List<TicketStatus>> call, @NonNull Throwable t) {
                Utils.log(t);
            }
        });

        //Set FAB to scanner
        floatingScanner.setOnClickListener(view -> openTicketScanActivity());
        // Set search function
        setupSearchView();

    }

    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.menu_activity_admin, menu);
        return true;
    }

    @Override
    public void onResume() {
        super.onResume();
        // Update the ticket that was being shown in the Ticket Detail View, in case the redemption state changed
        if (lastSelectedIndex < tickets.size()) {
            tickets.set(lastSelectedIndex,
                    TcaDb.getInstance(this).adminTicketDao().getByTicketId(tickets.get(lastSelectedIndex).getId()));
            ((TicketListAdapter) listView.getAdapter()).notifyDataSetChanged();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int i = item.getItemId();
        if (i == R.id.action_refresh) {
            refreshTicketList();
        }
        return true;
    }

    private void openTicketScanActivity() {
        Intent intent = new Intent(this, TicketScanActivity.class);
        intent.putExtra("eventId", eventID);
        startActivity(intent);
    }

    private void setupSearchView() {
        searchView.setSubmitButtonEnabled(true);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            public boolean onQueryTextSubmit(String queryText) {
                boolean searchSuccessful = search(queryText);

                if (!searchSuccessful) {
                    Utils.showToast(AdminDetailsActivity.this, getString(R.string.not_in_list));
                }

                return true;
            }

            public boolean onQueryTextChange(String queryText) {
                search(queryText);
                return true;
            }

            private boolean search(String queryText) {
                List<AdminTicket> foundTicketList = new ArrayList<>();
                for (AdminTicket foundTicket : tickets) {
                    if (foundTicket.getName().toLowerCase().contains(queryText.toLowerCase()) ||
                            foundTicket.getLrzId().toLowerCase().contains(queryText.toLowerCase())) {
                        foundTicketList.add(foundTicket);
                    }
                }
                // for an empty string all tickets are matched
                ticketListAdapter = new TicketListAdapter(foundTicketList, AdminDetailsActivity.this);
                ticketListAdapter.notifyDataSetChanged();
                listView.setAdapter(ticketListAdapter);

                return !foundTicketList.isEmpty();
            }
        });
    }

    private void refreshTicketList() {
        ticketsController.refreshTickets(eventID, list -> {
            tickets = list;
            ticketListAdapter = new TicketListAdapter(tickets, AdminDetailsActivity.this);
            listView.setAdapter(ticketListAdapter);
            ticketListAdapter.notifyDataSetChanged();
            updateTicketCounter();
        });
    }

    private void updateTicketCounter() {
        totalSaleView.setText(getString(R.string.sold_tickets, tickets.size(), totalTicketCount));
    }
}

File: app/src/main/java/de/tum/in/tca/ticketcheck/component/ticket/activity/AdminDetailsActivity.kt
package de.tum.`in`.tca.ticketcheck.component.ticket.activity

import androidx.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import com.google.android.material.snackbar.Snackbar
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.appcompat.widget.SearchView
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import de.tum.`in`.tca.ticketcheck.R
import de.tum.`in`.tca.ticketcheck.R.string.ticket
import de.tum.`in`.tca.ticketcheck.component.generic.activity.BaseActivity
import de.tum.`in`.tca.ticketcheck.component.ticket.EventsController
import de.tum.`in`.tca.ticketcheck.component.ticket.adapter.TicketsAdapter
import de.tum.`in`.tca.ticketcheck.component.ticket.fragment.TicketDetailsFragment
import de.tum.`in`.tca.ticketcheck.component.ticket.model.AdminTicket
import de.tum.`in`.tca.ticketcheck.component.ticket.model.Customer
import de.tum.`in`.tca.ticketcheck.component.ticket.model.TicketContingent
import de.tum.`in`.tca.ticketcheck.component.ticket.viewmodel.AdminDetailsViewModel
import de.tum.`in`.tca.ticketcheck.utils.Const
import de.tum.`in`.tca.ticketcheck.utils.observeNonNull
import kotlinx.android.synthetic.main.activity_admin.*

class AdminDetailsActivity : BaseActivity(
        R.layout.activity_admin
), TicketsAdapter.OnTicketSelectedListener, SwipeRefreshLayout.OnRefreshListener {

    private val ticketsAdapter: TicketsAdapter by lazy { TicketsAdapter(this) }

    private val snackbar: Snackbar by lazy {
        val view = findViewById<ViewGroup>(android.R.id.content)
        Snackbar.make(view.getChildAt(0), R.string.something_wrong, Snackbar.LENGTH_INDEFINITE)
    }

    private lateinit var searchItem: MenuItem

    private val eventID: Int by lazy { intent.getIntExtra(Const.EVENT_ID, 0) }

    private val viewModel: AdminDetailsViewModel by lazy {
        val factory = AdminDetailsViewModel.Factory(application, eventID)
        ViewModelProviders.of(this, factory).get(AdminDetailsViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        adminSwipeRefreshLayout.apply {
            setOnRefreshListener(this@AdminDetailsActivity)
            setColorSchemeResources(
                    R.color.color_primary,
                    R.color.tum_A100,
                    R.color.tum_A200)
        }

        val title = EventsController(this).getEventById(eventID).title
        supportActionBar?.title = title
        supportActionBar?.setSubtitle(R.string.loading)

        ticketsRecyclerView.apply {
            layoutManager = LinearLayoutManager(this@AdminDetailsActivity)
            setHasFixedSize(true)
            itemAnimator = DefaultItemAnimator()
            adapter = ticketsAdapter
        }

        viewModel.customers.observeNonNull(this) { updateTickets(it) }
        viewModel.ticketContingent.observeNonNull(this) { updateTicketCounter(it) }
        viewModel.error.observeNonNull(this) { showError(it) }

        refreshTicketsAndStats()

        scannerFab.setOnClickListener { openTicketScanActivity() }
    }

    override fun onTicketSelected(ticketIds: List<Int>, position: Int) {
        val fragment = TicketDetailsFragment.newInstance(ticketIds)
        fragment.show(supportFragmentManager, fragment.tag)
    }

    private fun openTicketScanActivity() {
        val intent = Intent(this, TicketScanActivity::class.java).apply {
            putExtra(Const.EVENT_ID, eventID)
        }
        startActivity(intent)
    }

    override fun onRefresh() {
        refreshTicketsAndStats()
    }

    private fun refreshTicketsAndStats() {
        adminSwipeRefreshLayout.isRefreshing = true
        viewModel.fetchTickets()
    }

    private fun updateTickets(customers: List<Customer>) {
        adminSwipeRefreshLayout.isRefreshing = false

        if (customers.isNotEmpty()) {
            ticketsRecyclerView.visibility = View.VISIBLE
            placeholderTextView.visibility = View.GONE
            ticketsAdapter.update(customers, eventID)
        } else {
            ticketsRecyclerView.visibility = View.GONE
            placeholderTextView.visibility = View.VISIBLE
        }
    }

    private fun updateTicketCounter(contingent: TicketContingent) {
        supportActionBar?.subtitle =
                getString(R.string.sold_tickets, contingent.sold, contingent.contingent)
    }

    private fun showError(show: Boolean) {
        adminSwipeRefreshLayout.isRefreshing = false
        if (show) {
            snackbar.show()
        } else {
            snackbar.dismiss()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_admin_details, menu)

        searchItem = menu.findItem(R.id.action_search)
        val searchView = searchItem.actionView as SearchView
        setupSearch(searchItem, searchView)

        return true
    }

    private fun setupSearch(searchItem: MenuItem, searchView: SearchView) {
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(queryText: String) = true

            override fun onQueryTextChange(query: String): Boolean {
                performSearch(query)
                return true
            }
        })

        searchView.setOnCloseListener {
            searchItem.collapseActionView()
            clearSearch()
            true
        }

        searchItem.setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
            override fun onMenuItemActionExpand(item: MenuItem?) = true

            override fun onMenuItemActionCollapse(item: MenuItem?): Boolean {
                clearSearch()
                return true
            }
        })
    }

    private fun performSearch(query: String) {
        viewModel.filter(query)
    }

    private fun clearSearch() {
        viewModel.filter(null)
    }

}
